libs/vkd3d/resource.c

```diff
@@ -2263,6 +2339,95 @@ HRESULT d3d12_resource_validate_desc(const D3D12_RESOURCE_DESC1 *desc, struct d3
             return E_INVALIDARG;
     }
 
+    /* Validate format cast list. */
+    if (num_castable_formats)
+    {
+        if (desc->Dimension == D3D12_RESOURCE_DIMENSION_BUFFER)
+        {
+            /* D3D12 runtime allows this, even if it is nonsensical. */
+            for (i = 0; i < num_castable_formats; i++)
+                if (castable_formats[i] != DXGI_FORMAT_UNKNOWN)
+                    return E_INVALIDARG;
+        }
+        else
+        {
+            const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+            VkFormatFeatureFlags required_flags = 0;
+            VkFormatFeatureFlags total_flags = 0;
+            VkFormatProperties format_properties;
+
+            format = vkd3d_format_from_d3d12_resource_desc(device, desc, 0);
```

Comment:
We might want to make this general, where we validate all resource descs, but it will mean constructing a temporary format cast list and iterating over that for the non-castable path.